### PR TITLE
NEW Archive admin for managing archived records

### PR DESCRIPTION
**Pull request by lukereative**
**Implements:** https://github.com/silverstripe/silverstripe-framework/issues/6202

**Sibling PRs:**
- https://github.com/silverstripe/silverstripe-framework/pull/8240
- https://github.com/silverstripe/silverstripe-versioned-admin/pull/29
- https://github.com/silverstripe/silverstripe-admin/pull/559

### CODE REVIEW

File: src/RestoreAction.php [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+     * @param Object $item
+     * @return array $message
+     */
+    public static function restore($item)
+    {
+        $isArchived = $item->isArchived();
+        $canRestoreToDraft = $item->canRestoreToDraft();
+
+        if (!$canRestoreToDraft) {
+            throw new ValidationException(
```
Comment (maxime-rainville) [inline src/RestoreAction.php]: Should have a 403 code.

File: src/RestoreAction.php [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+                _t(
+                    __CLASS__ . '.RESTORE_FALIURE_PERMISSION',
+                    'Insufficient permission to restore item'
+                )
+            );
+        }
+
+        if (!$isArchived) {
+            throw new ValidationException(
```
Comment (maxime-rainville) [inline src/RestoreAction.php]: Should return a 409 HTTP code.

File: src/Versioned.php [hunk 3]
```diff
@@ -1438,6 +1438,44 @@ public function canRevertToLive($member = null)
...
+    {
+        $owner = $this->owner;
+
+        // Skip if invoked by extendedCan()
+        if (func_num_args() > 4) {
+            return null;
+        }
+
+        // Can't revert if not archived
+        if (!$owner->isArchived()) {
```
Comment (maxime-rainville) [inline src/Versioned.php]: I don't think we should have this check here as it's not based on the Member's permission.

Basically, `canRestoreToDraft()` should return true if the provided Member has the right permission to restore the record, regardless of whatever the record is in a *restorable* state.

File: src/Versioned.php [hunk 4]
```diff
@@ -1438,6 +1438,44 @@ public function canRevertToLive($member = null)
         return $owner->canEdit($member);
     }
 
+    /**
+     * Check if the user can restore this record to draft
+     *
+     * @param Member $member
+     * @return bool
+     */
+    public function canRestoreToDraft($member = null)
```
Comment (maxime-rainville) [inline src/Versioned.php]: Probably wouldn't hurt to have a dedicated unit test for this one.

Comment (lukereative) [reply src/Versioned.php]: Yep, done

File: src/Versioned.php (hunk 3 above)
Comment (lukereative) [reply src/Versioned.php]: Makes sense, done

File: src/RestoreAction.php (hunk 2 above)
Comment (lukereative) [reply src/RestoreAction.php]: Done

File: src/RestoreAction.php (hunk 1 above)
Comment (lukereative) [reply src/RestoreAction.php]: Done

File: src/Versioned.php (hunk 3 above)
Comment (maxime-rainville) [reply src/Versioned.php]: Did you forget to push a commit? The archive check is still there.

**Pull request merged**